content/issue-2/applicative-parsing.md

```diff
@@ -0,0 +1,675 @@
+# Programming is hard, looking for better abstractions
+
+If you ever wrote a program bigger than a few lines of code you probably know that the bigger
+the program gets - the harder it's to make sure it does exactly what you meant due to how
+different parts can interact with each other. Over time programmers and computer scientists
+discovered multiple ways to manage this complexity. Programming went from physically wiring
+the instructions to writing them as mnemonic instructions, to abstractions with variables,
+labels, functions, and so on.
+
+
+[Functional Programming](https://en.wikipedia.org/wiki/Functional_programming?useskin=vector)
+is one of the paradigms for managing complexity with its main idea of writing code as a set of
+functions and using those functions to pass values around instead imperative style of steps
+to perform to mutate some data.
+
+Rust supports both functional and imperative approaches well and parts of its standard library
+are implemented with functional use in mind. Following examples show the difference between two
+approaches:
+
+Imperative approach describes exact steps program needs to take to calculate a sum of vector
+items:
+- initialize `sum` with 0
+- start at the first element of the array
+- multiply `i`th element of the array by two and add it to `sum`
+- proceed to the next element
+- repeat until the last element of the array
+
+```rust
+let xs = vec![1,2,3,4];
+let mut sum = 0;
+let mut i = 0;
+loop {
+    sum += xs[i] * 2;
+    i += 1;
+    if i == xs.len() {
+        break;
+    }
+}
```

Comment:
Well, yea, it's somewhat exaggerated to show explicit mutations and passing state around, etc. Using `for` and iterators makes it more like a traditional Rust, but that's mostly because of the language support for this particular task. I can go with manual toml/json parsing so language itself doesn't help much, but it's going to be a much bigger example. I guess I'll update the explanations here.

Reply:
how about using argument parsing as an example with a `getopts` like style for imperative and a teaser to the end of the tutorial for the functional approach

something like

```
fn parse() -> Args {
	let name = None;
	let answer = None;
	
	let args = std::env::args();
	while let (Some(arg), Some(value)) = (args.next(), args.next()){
		if arg == "--name" {
			name = Some(value.to_string());
		}
		if arg == "--answer" {
			answer = Some(value.parse::<i32>.unwrap()),
		}
	};

	Args {
		name: name.expect("'--name' not given"),
		answer: answer.expect("'--answer' not given"),
	}
}
```

also relates to your advantage that the functional approach plits business and boilerplate logic...

Reply:
This example is very good at showing horrors of imperative command line parsing but I think it's hard to make a fair and easy to understand functional counterpart. For that options are:
- dump all the required code from this article to make something that works - it will be a lot of code and I'm not sure if people are going to be interested in reading though without explanations and understanding how this works.
- show only compositions from this article or bpaf - while more readable - it's not quite apples to apples comparison. 

Calculating sum on the other hand is simple enough and people can relate to both versions. I could be mistaken though.

Reply:
I think the readerbase should have a vague idea of what functional programming is..
but it makes sense to say

> Imagine you want to create an arguemnt parser that parses your arguments into a struct like this:
> ```rust
> struct Args {
> 	name: String,
>   answer: i32,
> }
> ```
>
> If you want to write an argument parser purely imperatively, look at the horrors that entails, think about how to scale and maintain this:
> ```
> ... 
> ```
>
> What if you could use functional programming to just describe what you want:
>
> ```rust
> let name = Arg("--name").parse(|n| n.to_string())
> let answer = Arg("--answer").parse(|a| a.parse::<i32>)
> let Args = name
>	.zip(answer)
>	.map(|name, answer| Args { name, answer })
>	.run()
>	.unwrap()
> ```